### DISCUSSION TITLE
Job shop - constraining order of tasks

### DISCUSSION THREAD

**Discussion by edders51** [google/or-tools#3213] [CP-SAT questions] [answered]
Hi, 

I am using the Flexible Job shop with alternatives and transition times, minimally altered from the example.

My jobs consist of 3 tasks, and there are 2 machines (for example):

- Machine 0 - Can do tasks 0,2
- Machines 1 - Can do task 1.

```
    jobs = [[[(15, 0, 'L1')], [(10, 1, 'L1')], [(15, 0, 'L1')]],
              [[(15, 0, 'L2')], [(30, 1, 'L2')], [(15, 0, 'L2')]]]
```

What i would like to do, but am struggling with, is to stipulate that whenever machine 1 does task 1, machine 0 must finish the associated task 2, and the task 0 of the next job on machine 1 before that machine 1 can start it's next task 1 (e.g. for the next job).

The reason i am using separate machines, and not the same, is in my real example task 1 can be done by multiple machines, and is a longer task than tasks 0 & 2, which can only be done on one machine.

I have a precedent such that tasks must be completed in order, but this does not prevent a subsequent job from commencing.
E.G i want:
Machine 0 --- J0_T0====J0_T2=**J1_T0**====**J1_T2**
Machine 1 ---====J0_T1========**J1_T1**
Or:
Machine 0 --- J0_T0=**J1_T0**=J0_T2=**J1_T2**
Machine 1 ---=====J0_T1==========
Machine 2 ---=========**J1_T1**======

What happens:
Machine 0 --- J0_T0=**J1_T0**=J0_T2=**J1_T2**
Machine 1 --- ====J0_T1=**J1_T1**=====

Any pointers gratefully received, I am new to OR.

Cheers,
Edwin

**Comment by lperron** [comment]
are these just precedences between tasks ?

**Reply by edders51** [reply]
I think it is more than that.
The bottom wrong example follows the precedence between tasks but is not what i am looking for.

I want the precedence to depend on the order that Machine 1 does jobs. I want machine 0 to complete task 2 for any job on machine 1, before starting the next task that will go to machine 1. In my "Or" example, we can start job 1 before job 0 has completed its final task.

Not sure i am explaining this very well...

**Comment by lperron** [comment, marked as answer]
I would suggest a set of booleans b_i_j <=> task_i before task_j. And then write the implications between all these Booleans.
